### PR TITLE
feat(failover): Sunspot bad-node patterns + machine-agnostic --auto-retry driver (validated on Sunspot)

### PR DESCRIPTION
**Pull request by saforem2**
## Summary

Adds Sunspot support to the failover stack and generalizes the `--auto-retry`
integration test driver — then **validates it end-to-end on a live 4-node
Sunspot allocation (7/7 scenarios pass)**, the first time the `ezpz launch
--auto-retry` swap-and-retry loop has run on real PALS hardware. The live run
also caught a **real classifier bug** (below).

## What's here

**1. `src/ezpz/failover/patterns/sunspot.py` (new).** Sunspot is Aurora's
test/dev twin (same PVC + PALS) but a different hostname suffix
(`.hsn.cm.sunspot.alcf.anl.gov`, optional `-hsnN` on the node token). Before
this, `scrape_bad_nodes` on Sunspot imported a nonexistent module → `[]` →
`--auto-retry` could only blind-rotate, never a named swap. Mirrors `aurora.py`.

**2. `fix(failover)`: retry when a nonzero rank exit surfaces as rc=143.**
Found by the live run. On PALS, a rank's nonzero app exit tears the job down
with SIGTERM → rc=143 (`_WALLTIME_RC`) + `rank N exited with code K`. That
wasn't a crash pattern, so a genuine crash was misread as a clean walltime kill
and NOT retried. Added `rank \d+ exited with code [1-9]` to the Python + bash
crash patterns (the `[1-9]` bound excludes PALS's clean `code 0` lines).

**3. Generalized PBS driver → `test_launch_auto_retry.pbs`.** `git mv` from
`_aurora.pbs`; runtime GPUs/node, per-machine qsub flags, preflight scraper
self-check, race-free marker-file fakes, `rc!=0` assertions, per-job log path.

## Tests
+11 Sunspot scrape tests, +2 classifier tests, +2 bash tests. **149 python +
23 bash failover tests pass; ruff clean.**

## Live Sunspot validation (job 12471669, 4 nodes)
Preflight named the real `-hsn0` host; **A–G all PASS (7/7)** — including
scenario **C's real named swap** driven by the new `sunspot.py` pattern.

## Test plan
- [x] Local: pytest (149) + bash (23) + ruff clean
- [x] Live Sunspot: 7/7 on a real 4-node allocation

## Summary by Sourcery

Add Sunspot-specific failover patterns and extend crash classification to correctly retry PALS nonzero-rank exits surfaced as rc=143.

New Features:
- Introduce Sunspot-specific bad-node scraping patterns and hostname normalization registered under the Sunspot machine key.
- Generalize the launch auto-retry PBS driver script to be machine-agnostic for running integration scenarios beyond Aurora.

Bug Fixes:
- Treat PALS logs with nonzero rank exit codes surfaced as rc=143 as bad-node crashes rather than clean walltime exits, while preserving clean handling of rank exit code 0 cases.

Tests:
- Add Sunspot scraping and hostname normalization unit tests to validate bad-node detection on Sunspot logs.
- Add regression tests in Python and bash to cover rc=143 classification for nonzero and zero rank exit codes in the auto-retry flow.

### CODE REVIEW

File: src/ezpz/launch_autoretry.py [hunk 1]
```diff
@@ -69,6 +69,17 @@
...
+    # PALS, one rank's `exit 1` tears the job down with SIGTERM and the
+    # aggregate surfaces as bash 143 (== _WALLTIME_RC) plus a
+    # `<host>: rank N exited with code K` line (K != 0). Without this
+    # pattern a genuine crash is indistinguishable from a clean walltime
+    # kill (which only ever emits `rank N died from signal {11,15}`,
+    # stripped as an innocent cascade below) and the loop would NOT retry
+    # it. Confirmed on a live Sunspot run (job 12471663): scenarios B/D/E
+    # all stopped after 1-2 attempts as "walltime" for a plain `exit 1`.
+    # `[1-9]` excludes the clean `exited with code 0` PALS prints per rank.
+    r"|rank \d+ exited with code [1-9]"
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/launch_autoretry.py]: Fixed in 84c5b83 — widened to `rank \d+ exited with code [1-9][0-9]*`. You are right that this matters: code 137 (128+SIGKILL, i.e. an OOM-kill) is exactly the kind of nonzero exit we most want to retry, and the single-digit form silently missed it.

File: src/ezpz/bin/failover.sh [hunk 2]
```diff
@@ -311,7 +311,7 @@ failover_run() {
             # (real shepherd kill), `died from signal 6` (SIGABRT
             # from a real assert), etc. still count.
             crash_lines=$(grep -vE "rank [0-9]+ died from signal (11|15)" "$logf" 2>/dev/null \
-                | grep -cE "RuntimeError: \[.*gloo.*\] Connection closed by peer|RuntimeError: \[.*gloo.*\] Timed out waiting|OutOfMemoryError|UR_RESULT_ERROR_OUT_OF_RESOURCES|died from signal|EOFError: No data left in file")
+                | grep -cE "RuntimeError: \[.*gloo.*\] Connection closed by peer|RuntimeError: \[.*gloo.*\] Timed out waiting|OutOfMemoryError|UR_RESULT_ERROR_OUT_OF_RESOURCES|died from signal|rank [0-9]+ exited with code [1-9]|EOFError: No data left in file")
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/bin/failover.sh]: Fixed in 84c5b83 — both `grep -cE` lists in failover.sh now use `rank [0-9]+ exited with code [1-9][0-9]*`, matching the Python side.

File: src/ezpz/failover/patterns/sunspot.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+# PBS hostfile entries and PALS shepherd lines use the
+# `.hsn.cm.sunspot.alcf.anl.gov` form, with the HSN node token optionally
+# carrying a `-hsnN` suffix. The `.hostmgmtNNNN.cm.sunspot...` management
+# form maps 1:1 to the HSN interface and is safe to rewrite. Any OTHER
+# suffix on an `x...n0.` host is something we haven't seen and shouldn't
+# speculatively rewrite — return None (drop) rather than risk tagging a
+# wrong node.
+# ---------------------------------------------------------------------------
+_SUNSPOT_HSN_RX = re.compile(
+    r"^x\d+c\d+s\d+b\d+n\d+(?:-hsn\d+)?\.hsn\.cm\.sunspot\.alcf\.anl\.gov$"
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/failover/patterns/sunspot.py]: Good catch — fixed in 84c5b83. `normalize_sunspot_hostname` now strips the `-hsnN` token so both `x..n0-hsn0.hsn...` and `x..n0.hsn...` canonicalize to the single suffix-less form. Added a same-node-two-forms dedup test.

File: tests/test_launch_autoretry.py [hunk 4]
```diff
@@ -270,6 +270,44 @@ def test_walltime_with_cascade_and_real_death_is_bad_node(
...
+    def test_pals_nonzero_exit_surfaces_as_143_is_bad_node(self, tmp_path):
+        """REGRESSION (live Sunspot job 12471663): on Aurora/Sunspot PALS a
+        rank's nonzero application exit tears the job down with SIGTERM, so
+        the aggregate shell rc is 143 (== _WALLTIME_RC) and the log shows
+        `<host>: rank N exited with code K` (K != 0) alongside innocent
+        `rank M died from signal 15` cascade lines. Before the
+        `rank N exited with code [1-9]` crash pattern, this classified as a
+        clean WALLTIME and the loop did NOT retry a genuine crash. It must
+        be a bad-node retry.
+        """
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_launch_autoretry.py:282]: Added in 84c5b83 — `test_pals_mixed_exit_codes_is_bad_node` (rc=143 with both code 0 and code 1 → BAD_NODE_BLIND), plus `test_pals_multidigit_exit_code_is_bad_node` (code 137) guarding the `[1-9][0-9]*` widening.

File: src/ezpz/bin/test_launch_auto_retry.pbs [hunk 5]
```diff
@@ -21,17 +26,33 @@
...
 #
-# Submit with:
-#   cd ~/ezpz  # (or wherever your checkout is)
-#   qsub src/ezpz/bin/test_launch_auto_retry_aurora.pbs
+# The `#PBS` directives above default to Aurora, but qsub command-line
+# options OVERRIDE script directives — so submit with the per-machine
+# flags below (nothing inside the script is Aurora-specific):
+#
+#   # Aurora:
+#   qsub src/ezpz/bin/test_launch_auto_retry.pbs
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/bin/test_launch_auto_retry.pbs:34]: Fixed in 84c5b83 — docs/cli/launch/index.md now points qsub at `test_launch_auto_retry.pbs` (the renamed driver) with per-machine submit flags (Aurora/Sunspot), instead of the deleted `_aurora.pbs` path.
